### ISSUE TITLE
ordering of columns in mlr3

### ISSUE THREAD

**Issue by berndbischl** [mlr-org/mlr3#365]
Does mlr3 somehow guarantee / ensure that column orders do not change?
Are you @mllg aware of this constraint?

a) We found a problem with at least xgboost of order changes. so if xgboost is passed data during training and predict, and that data is "legal" in principle, but the column order has changed during training and predict, xgboost fails. and there might be other learners where this happens.

b) I found this when I tested mlr3pipelines with imputation. This changes the order of cols. Then I ran into this bug.

c) So this is certainly also a pipelines issue. but @mb706 told me he asked you about this, and you said "column order does not matter / can be arbitrary" so at least I wanted to raise this here too. 

As Martin, when he changes data in preproc, often changes data by deleting the col that needs to be changed,  then "cbinds" the changed col to end of the dataset, we now potentially face this problem a lot. And this can completely break stuff

**Comment by mllg** [comment]
> Does mlr3 somehow guarantee / ensure that column orders do not change?

Nope. From `?DataBackend`:
> Rows are guaranteed to be returned in the same order as `rows`, columns may be returned in an arbitrary order.

Reasoning: Virtual backends (e.g., cbind of two backends) do not support an ordering of columns.


> a) We found a problem with at least xgboost of order changes. so if xgboost is passed data during training and predict, and that data is "legal" in principle, but the column order has changed during training and predict, xgboost fails. and there might be other learners where this happens.

That is probably because the target column is not part of the XGB matrix (it's more like an attribute attached to the matrix). Otherwise, this would always emerge during predict if the target column is not the first/last column. Thus, I doubt that there are more learners which are affected by this, but I'll write an autotest for this.


> As Martin, when he changes data in preproc, often changes data by deleting the col that needs to be changed, then "cbinds" the changed col to end of the dataset, we now potentially face this problem a lot. And this can completely break stuff

Accessing data frame columns by number is clearly something you should not do. Can you name more situations / packages where this breaks something?
I could ensure an ordering in `task$data()`, but I would rather not if not absolutely necessary.

**Comment by berndbischl** [comment]
> That is probably because the target column is not part of the XGB matrix (it's more like an attribute attached to the matrix). Otherwise, this would always emerge during predict if the target column is not the first/last column. Thus, I doubt that there are more learners which are affected by this, but I'll write an autotest for this.

it is not. it is really the ordering. i am 99% sure, and started to debug this yesterday by really inspecting what data is passed to xgboost. we can also try to build a reproducing example.
and if you are wrong here: how could you ever know that other learners do not require the same thing? if "bake into" mlr3 or pipelines "reordering" of cols, stuff will then become very instable if we dont do something about this.
basically my FIRST SUPER SIMPLE usecase with xgboost and pipelines failed beacause of this.

**Comment by berndbischl** [comment]
> Accessing data frame columns by number is clearly something you should not do.

huh? where did i say we access by numeric indices?
@mb706 can you pls link to the pipeline code where we can see where the reordering happens?

**Comment by mllg** [comment]
You misunderstood. xgboost relies on the order. Usually, you don't change the order of the columns between train and predict. But this happens automatically if you pass a test data set w/o target column (with the exception that the target is in the last column). But because xgboost has a native (tabular) data format where the target and features are separated, there usually is no reordering.

**Comment by berndbischl** [comment]
> Usually, you don't change the order of the columns between train and predict. But this happens automatically if you pass a test data set w/o target column (with the exception that the target is in the last column).

but thats not my point? you misunderstand. martin DOES reorder the columns. during train and predict. i can see that. every impute operation can do that. because of the way you 2 guys agreed who to do that (by cbinding changed stuff instead of changing in place). and then xgboost breaks. 

i will post an example later.

**Comment by mllg** [comment]
I know that xgboost breaks, that's easily reproducible. I also know that we reorder the columns. My point is that the learners should not rely on the order of the columns.

I've added an autotest to detect such learners in 4ac5f349bab076c496d6e2d3de0c4c4c0b926caf. This breaks xgboost and svm (matrix interface) in mlr3learners.

**Comment by mllg** [comment]
https://github.com/mlr-org/mlr3learners/commit/86ff8e177e3c9a76bfc2753e2af7c1a04ad82efe repairs xgboost and svm. Note that this is probably more like a temporary fix.

**Comment by mllg** [comment]
> because of the way you 2 guys agreed on how to implement task-data-changes (by cbinding changed stuff instead of changing in place).

How would you change in place for a remote data base backend?

**Comment by berndbischl** [comment]
> My point is that the learners should not rely on the order of the columns.

you are saying we should stabilize the interface code in mlr3learners? as you are already apperently starting to do? 
because then i at least understand your point of view - and its probably (?) the best route? 
a bit annoying then to do it in each learner.
but 
a) if we document this
AND
b) have a test
AND
c) dont know what to do alternatively

then lets do that?  agreed?

**Comment by berndbischl** [comment]
> How would you change in place for a remote data base backend?

i know. and see my answer above. BUT i want to say this for completeness sake:

your quote above you have used now as an argument MANY TIMES for "problematic scenarios" / why we cannot use certain approaches.
it's not really a criticism, but an honest observation. 
less then a few of our users are in such a scenario, very few times when they use mlr. 
just saying. 
providing the features of mlr3db is super nice. but it makes sense to consider the price of this.

(apparently, we can fix this problem, here, now, so this is a general comment, but i had that feeling now a couple of times....)

**Comment by mllg** [comment]
> a) if we document this

In the book now.

> b) have a test

Is an autotest now.

> c) dont know what to do alternatively

I could be more conservative with column reordering, but this probably won't solve all problems.

**Comment by mohammadreza-sheykhmousa** [comment]
Dear mlr3 team,

I was tring to pass `pipeline_robustify` to preprocess my task first but faced with this error: 
```r
>         lrn = lrn("classif.featureless")
>         gr = pipeline_robustify(tsk_clf, lrn) %>>% po("imputemode") %>>% po("learner", lrn)
>         ede = resample(tsk_clf, GraphLearner$new(gr), rsmp("cv", folds = 2))
INFO  [07:38:52.943] Applying learner 'removeconstants.imputemode.classif.featureless' on task 'df.tr' (iter 1/2) 
Error in .__Task__col_roles(self = self, private = private, super = super,  : 
  Assertion on 'names of col_roles' failed: Must be a permutation of set {feature,target,name,order,stratum,group,weight}.
```
I know has to do with `col_roles`

```r
> names(tsk_clf$col_roles)
[1] "feature"     "target"      "label"       "order"       "group"       "weight"      "coordinates"
```
shall I set a col_role here and add `stratum` here? If so, how?

```r
> tsk_clf$strata = TRUE
Error: Field/Binding is read-only
```

**Comment by mohammadreza-sheykhmousa** [comment]
Update!
I find the source of the error as mentioned in the `traceback` it was related to `future.apply`. So I run the code without `future.apply` and works fine. but the question here is where and when one should use `future.apply` ? bc most of the time for me doesn't parallelized the process. I'd be more than happy to hear your suggestions.
Best, Mo

**Comment by mllg** [comment]
> [1] "feature"     "target"      "label"       "order"       "group"       "weight"      "coordinates"

Where do the extra column roles come from? "label" and "coordinates" are not in mlr3.

Can you provide a [reprex](https://gist.github.com/hadley/270442)?

**Comment by mohammadreza-sheykhmousa** [comment]
Hi @mllg this is `reprex`  point is now somehow problem solved even with `future.apply` and my problem is now with `NA's` which i was hoping `pipeline_robustify` would have solve it at the first place which as you will see wasn't apparently the case.

``` r
library("reprex")
library("mlr3verse")

library("bbotk")
library("ggplot2")
library("mltools")
library("data.table")
library("mlr3fselect")
library("FSelectorRcpp")
library("future")
library("future.apply")
library("magrittr")

df <- read.csv('/home/msheykhmousa/Downloads/training_samples_overlayed.csv',header = TRUE)
names(df)
#>  [1] "X"                              "lucas"                         
#>  [3] "survey_date"                    "lc_class"                      
#>  [5] "tile_id"                        "confidence"                    
#>  [7] "geometry"                       "overlay_id"                    
#>  [9] "landsat_ard_winter_blue_p50"    "landsat_ard_winter_green_p25"  
#>  .....
#> [95] "landsat_ard_fall_swir2_p50"     "landsat_ard_winter_thermal_p25"
#> [97] "landsat_ard_fall_thermal_p25"   "dtm_elevation"                 
#> [99] "dtm_slope"
smp_size <- floor(0.5 * nrow(df))
set.seed(123)
train_ind <- sample(seq_len(nrow(df)), size = smp_size)
df.tr <- df[train_ind, ]
df.ts <- df[-train_ind, ]
folds = 2
n_evals = 2
target.variable = "lc_class"
cv3 = rsmp("repeated_cv", folds = folds)

df.tr$lc_class <- as.factor(df.tr[,"lc_class"])
id = deparse(substitute(df.tr))
tsk_clf <- mlr3::TaskClassif$new(id = id, backend = df.tr, target = target.variable)
names(tsk_clf$col_roles)
#> [1] "feature" "target"  "name"    "order"   "stratum" "group"   "weight"
lrn = lrn("classif.featureless")
gr = pipeline_robustify(tsk_clf, lrn) %>>% po("imputemode") %>>% po("learner", lrn)
future::plan(multisession, globals = TRUE,
             workers = 32)
ede = resample(tsk_clf, GraphLearner$new(gr), rsmp("cv", folds = 2))
#> INFO  [12:20:18.796] Applying learner 'removeconstants.imputemode.classif.featureless' on task 'df.tr' (iter 2/2) 
#> INFO  [12:20:19.148] Applying learner 'removeconstants.imputemode.classif.featureless' on task 'df.tr' (iter 1/2)
tsk_clasif1 = ede$task$clone()
names(tsk_clasif1$col_roles)
#> [1] "feature" "target"  "name"    "order"   "stratum" "group"   "weight"
ranger_lrn = lrn("classif.ranger", predict_type = "response",importance ="permutation")
# model = level0 %>>% ranger_lrn
ps_ranger = ParamSet$new(
  list(
    ParamInt$new("mtry", lower = 1L, upper = 5L),
    ParamDbl$new("sample.fraction", lower = 0.5, upper = 1),
    ParamInt$new("num.trees", lower = 50L, upper = 500L),
    # ParamInt$new("classif.ranger.num.threads",  31, 32),
    ParamFct$new("importance", "permutation")
  ))

at = AutoTuner$new(
  learner = ranger_lrn,
  resampling = cv3,
  measure = msr("classif.acc"),
  search_space = ps_ranger,
  terminator = trm("evals", n_evals = n_evals), 
  tuner = tnr("random_search")
)
at$store_tuning_instance = TRUE

requireNamespace("lgr")
logger = lgr::get_logger("mlr3")
logger$set_threshold("trace")
lgr::get_logger("mlr3")$set_threshold("warn")
lgr::get_logger("mlr3")$set_threshold("debug")
at$train(tsk_clasif1)
#> DEBUG [12:20:21.324] Skip subsetting of task 'df.tr' 
#> DEBUG [12:20:21.373] Calling train method of Learner 'classif.ranger.tuned' on task 'df.tr' with 379 observations {learner: <AutoTuner/Learner/R6>}
#> INFO  [12:20:21.467] Starting to optimize 4 parameter(s) with '<OptimizerRandomSearch>' and '<TerminatorEvals>' 
#> INFO  [12:20:21.530] Evaluating 1 configuration(s) 
#> INFO  [12:20:21.580] Benchmark with 20 resampling iterations 
#> DEBUG [12:20:21.581] Running benchmark() asynchronously with 20 iterations 
#> INFO  [12:20:21.804] Applying learner 'classif.ranger' on task 'df.tr' (iter 16/20) 
#> DEBUG [12:20:21.829] Subsetting task 'df.tr' to 190 rows {task: <TaskClassif/TaskSupervised/Task/R6>, row_ids: [1, 4, 5, 6, 7, 9, 10, 11, 13, 17, 19, 20, 22, 25, 26, 28, 31, 34, 36, 37, 41, 42, 44, 45, 54, 57, 59, 62, 64, 66..]}
#> DEBUG [12:20:21.834] Calling train method of Learner 'classif.ranger' on task 'df.tr' with 190 observations {learner: <LearnerClassifRanger/LearnerClassif/Learner/R6>}
#> Error: Missing data in columns: landsat_ard_fall_blue_p25, landsat_ard_fall_blue_p50, landsat_ard_fall_blue_p75, landsat_ard_fall_count, landsat_ard_fall_green_p25, landsat_ard_fall_green_p50, landsat_ard_fall_green_p75, landsat_ard_fall_nir_p25, landsat_ard_fall_nir_p50, landsat_ard_fall_nir_p75, landsat_ard_fall_red_p25, landsat_ard_fall_red_p50, landsat_ard_fall_red_p75, landsat_ard_fall_swir1_p25, landsat_ard_fall_swir1_p50, landsat_ard_fall_swir1_p75, landsat_ard_fall_swir2_p25, landsat_ard_fall_swir2_p50, landsat_ard_fall_swir2_p75, landsat_ard_fall_thermal_p25, landsat_ard_fall_thermal_p50, landsat_ard_fall_thermal_p75, landsat_ard_winter_blue_p25, landsat_ard_winter_blue_p50, landsat_ard_winter_blue_p75, landsat_ard_winter_count, landsat_ard_winter_green_p25, landsat_ard_winter_green_p50, landsat_ard_winter_green_p75, landsat_ard_winter_nir_p25, landsat_ard_winter_nir_p50, landsat_ard_winter_nir_p75, landsat_ard_winter_red_p25, landsat_ard_winter_red_p50, landsat_ard_winter_red_p75, landsat_ard_winter_swir1_p25, landsat_ard_winter_swir1_p50, landsat_ard_winter_swir1_p75, landsat_ard_winter_swir2_p25, landsat_ard_winter_swir2_p50, landsat_ard_winter_swir2_p75, landsat_ard_winter_thermal_p25, landsat_ard_winter_thermal_p50, landsat_ard_winter_thermal_p75.

at$learner$train(tsk_clasif1)
#> DEBUG [12:20:28.747] Skip subsetting of task 'df.tr' 
#> DEBUG [12:20:28.750] Calling train method of Learner 'classif.ranger' on task 'df.tr' with 379 observations {learner: <LearnerClassifRanger/LearnerClassif/Learner/R6>}
#> Error: Missing data in columns: landsat_ard_fall_blue_p25, landsat_ard_fall_blue_p50, landsat_ard_fall_blue_p75, landsat_ard_fall_count, landsat_ard_fall_green_p25, landsat_ard_fall_green_p50, landsat_ard_fall_green_p75, landsat_ard_fall_nir_p25, landsat_ard_fall_nir_p50, landsat_ard_fall_nir_p75, landsat_ard_fall_red_p25, landsat_ard_fall_red_p50, landsat_ard_fall_red_p75, landsat_ard_fall_swir1_p25, landsat_ard_fall_swir1_p50, landsat_ard_fall_swir1_p75, landsat_ard_fall_swir2_p25, landsat_ard_fall_swir2_p50, landsat_ard_fall_swir2_p75, landsat_ard_fall_thermal_p25, landsat_ard_fall_thermal_p50, landsat_ard_fall_thermal_p75, landsat_ard_summer_blue_p25, landsat_ard_summer_blue_p50, landsat_ard_winter_blue_p25, landsat_ard_winter_blue_p50, landsat_ard_winter_blue_p75, landsat_ard_winter_count, landsat_ard_winter_green_p25, landsat_ard_winter_green_p50, landsat_ard_winter_green_p75, landsat_ard_winter_nir_p25, landsat_ard_winter_nir_p50, landsat_ard_winter_nir_p75, landsat_ard_winter_red_p25, landsat_ard_winter_red_p50, landsat_ard_winter_red_p75, landsat_ard_winter_swir1_p25, landsat_ard_winter_swir1_p50, landsat_ard_winter_swir1_p75, landsat_ard_winter_swir2_p25, landsat_ard_winter_swir2_p50, landsat_ard_winter_swir2_p75, landsat_ard_winter_thermal_p25, landsat_ard_winter_thermal_p50, landsat_ard_winter_thermal_p75.
tr.mdl = at$learner
train.model = tr.mdl$predict_newdata
```

<sup>Created on 2020-09-29 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

<details>

<summary>Session info</summary>

``` r
devtools::session_info()
#> ─ Session info ───────────────────────────────────────────────────────────────
#>  setting  value                       
#>  version  R version 4.0.2 (2020-06-22)
#>  os       Ubuntu 20.04.1 LTS          
#>  system   x86_64, linux-gnu           
#>  ui       X11                         
#>  language (EN)                        
#>  collate  en_US.UTF-8                 
#>  ctype    en_US.UTF-8                 
#>  tz       Europe/Amsterdam            
#>  date     2020-09-29                  
#> 
#> ─ Packages ───────────────────────────────────────────────────────────────────
#>  package       * version    date       lib
#>  assertthat      0.2.1      2019-03-21 [1]
#>  backports       1.1.10     2020-09-15 [1]
#>  bbotk         * 0.2.1      2020-09-05 [1]
#>  callr           3.4.4      2020-09-07 [1]
#>  checkmate       2.0.0      2020-02-06 [1]
#>  cli             2.0.2      2020-02-28 [1]
#>  codetools       0.2-16     2018-12-24 [3]
#>  colorspace      1.4-1      2019-03-18 [1]
#>  crayon          1.3.4      2017-09-16 [1]
#>  data.table    * 1.13.0     2020-07-24 [1]
#>  desc            1.2.0      2018-05-01 [1]
#>  devtools        2.3.1      2020-07-21 [1]
#>  digest          0.6.25     2020-02-23 [1]
#>  dplyr           1.0.2      2020-08-18 [1]
#>  ellipsis        0.3.1      2020-05-15 [1]
#>  evaluate        0.14       2019-05-28 [1]
#>  fansi           0.4.1      2020-01-08 [1]
#>  foreach         1.5.0      2020-03-30 [1]
#>  fs              1.5.0      2020-07-31 [1]
#>  FSelectorRcpp * 0.3.3      2020-01-24 [1]
#>  future        * 1.19.1     2020-09-22 [1]
#>  future.apply  * 1.6.0      2020-07-01 [1]
#>  generics        0.0.2      2018-11-29 [1]
#>  ggplot2       * 3.3.2      2020-06-19 [1]
#>  globals         0.13.0     2020-09-17 [1]
#>  glue            1.4.2      2020-08-27 [1]
#>  gtable          0.3.0      2019-03-25 [1]
#>  highr           0.8        2019-03-20 [1]
#>  htmltools       0.5.0      2020-06-16 [1]
#>  iterators       1.0.12     2019-07-26 [1]
#>  knitr           1.29       2020-06-23 [1]
#>  lattice         0.20-41    2020-04-02 [1]
#>  lgr             0.3.4      2020-03-20 [1]
#>  lifecycle       0.2.0      2020-03-06 [1]
#>  listenv         0.8.0      2019-12-05 [1]
#>  magrittr      * 1.5        2014-11-22 [1]
#>  Matrix          1.2-18     2019-11-27 [1]
#>  memoise         1.1.0      2017-04-21 [1]
#>  mlr3          * 0.6.0-9000 2020-09-24 [1]
#>  mlr3filters   * 0.3.0.9000 2020-09-11 [1]
#>  mlr3fselect   * 0.2.0      2020-08-23 [1]
#>  mlr3learners  * 0.3.0      2020-08-29 [1]
#>  mlr3measures    0.2.0      2020-06-27 [1]
#>  mlr3misc        0.5.0      2020-08-13 [1]
#>  mlr3pipelines * 0.2.1-9000 2020-09-25 [1]
#>  mlr3tuning    * 0.2.0      2020-07-28 [1]
#>  mlr3verse     * 0.1.3      2020-07-06 [1]
#>  mlr3viz       * 0.2.0      2020-08-07 [1]
#>  mltools       * 0.3.5      2018-05-12 [1]
#>  munsell         0.5.0      2018-06-12 [1]
#>  paradox       * 0.4.0      2020-07-21 [1]
#>  pillar          1.4.6      2020-07-10 [1]
#>  pkgbuild        1.1.0      2020-07-13 [1]
#>  pkgconfig       2.0.3      2019-09-22 [1]
#>  pkgload         1.1.0      2020-05-29 [2]
#>  prettyunits     1.1.1      2020-01-24 [1]
#>  processx        3.4.4      2020-09-03 [1]
#>  ps              1.3.4      2020-08-11 [1]
#>  purrr           0.3.4      2020-04-17 [1]
#>  R6              2.4.1      2019-11-12 [1]
#>  ranger          0.12.1     2020-01-10 [1]
#>  Rcpp            1.0.5      2020-07-06 [1]
#>  remotes         2.2.0      2020-07-21 [1]
#>  reprex        * 0.3.0      2019-05-16 [1]
#>  rlang           0.4.7      2020-07-09 [1]
#>  rmarkdown       2.3        2020-06-18 [1]
#>  rprojroot       1.3-2      2018-01-03 [1]
#>  scales          1.1.1      2020-05-11 [1]
#>  sessioninfo     1.1.1      2018-11-05 [1]
#>  stringi         1.4.6      2020-02-17 [1]
#>  stringr         1.4.0      2019-02-10 [1]
#>  testthat        2.3.2      2020-03-02 [2]
#>  tibble          3.0.3      2020-07-10 [1]
#>  tidyselect      1.1.0      2020-05-11 [1]
#>  usethis         1.6.1      2020-04-29 [1]
#>  uuid            0.1-4      2020-02-26 [1]
#>  vctrs           0.3.4      2020-08-29 [1]
#>  withr           2.3.0      2020-09-22 [1]
#>  xfun            0.16       2020-07-24 [1]
#>  yaml            2.2.1      2020-02-01 [1]
#>  source                                
#>  CRAN (R 4.0.2)                        
#>  CRAN (R 4.0.2)                        
#>  CRAN (R 4.0.2)                        
#>  CRAN (R 4.0.2)                        
#>  CRAN (R 4.0.2)                        
#>  CRAN (R 4.0.2)                        
#>  CRAN (R 4.0.0)                        
#>  CRAN (R 4.0.2)                        
#>  CRAN (R 4.0.2)                        
#>  CRAN (R 4.0.2)                        
#>  CRAN (R 4.0.2)                        
#>  CRAN (R 4.0.2)                        
#>  CRAN (R 4.0.2)                        
#>  CRAN (R 4.0.2)                        
#>  CRAN (R 4.0.2)                        
#>  CRAN (R 4.0.2)                        
#>  CRAN (R 4.0.2)                        
#>  CRAN (R 4.0.2)                        
#>  CRAN (R 4.0.2)                        
#>  CRAN (R 4.0.2)                        
#>  CRAN (R 4.0.2)                        
#>  CRAN (R 4.0.2)                        
#>  CRAN (R 4.0.2)                        
#>  CRAN (R 4.0.2)                        
#>  CRAN (R 4.0.2)                        
#>  CRAN (R 4.0.2)                        
#>  CRAN (R 4.0.2)                        
#>  CRAN (R 4.0.2)                        
#>  CRAN (R 4.0.2)                        
#>  CRAN (R 4.0.2)                        
#>  CRAN (R 4.0.2)                        
#>  CRAN (R 4.0.2)                        
#>  CRAN (R 4.0.2)                        
#>  CRAN (R 4.0.2)                        
#>  CRAN (R 4.0.2)                        
#>  CRAN (R 4.0.2)                        
#>  CRAN (R 4.0.2)                        
#>  CRAN (R 4.0.2)                        
#>  Github (mlr-org/mlr3@e9d500e)         
#>  Github (mlr-org/mlr3filters@faf79ad)  
#>  CRAN (R 4.0.2)                        
#>  CRAN (R 4.0.2)                        
#>  CRAN (R 4.0.2)                        
#>  CRAN (R 4.0.2)                        
#>  Github (mlr-org/mlr3pipelines@fae2715)
#>  CRAN (R 4.0.2)                        
#>  CRAN (R 4.0.2)                        
#>  CRAN (R 4.0.2)                        
#>  CRAN (R 4.0.2)                        
#>  CRAN (R 4.0.2)                        
#>  CRAN (R 4.0.2)                        
#>  CRAN (R 4.0.2)                        
#>  CRAN (R 4.0.2)                        
#>  CRAN (R 4.0.2)                        
#>  CRAN (R 4.0.0)                        
#>  CRAN (R 4.0.2)                        
#>  CRAN (R 4.0.2)                        
#>  CRAN (R 4.0.2)                        
#>  CRAN (R 4.0.2)                        
#>  CRAN (R 4.0.2)                        
#>  CRAN (R 4.0.2)                        
#>  CRAN (R 4.0.2)                        
#>  CRAN (R 4.0.2)                        
#>  CRAN (R 4.0.2)                        
#>  CRAN (R 4.0.2)                        
#>  CRAN (R 4.0.2)                        
#>  CRAN (R 4.0.2)                        
#>  CRAN (R 4.0.2)                        
#>  CRAN (R 4.0.2)                        
#>  CRAN (R 4.0.2)                        
#>  CRAN (R 4.0.2)                        
#>  CRAN (R 4.0.0)                        
#>  CRAN (R 4.0.2)                        
#>  CRAN (R 4.0.2)                        
#>  CRAN (R 4.0.2)                        
#>  CRAN (R 4.0.2)                        
#>  CRAN (R 4.0.2)                        
#>  CRAN (R 4.0.2)                        
#>  CRAN (R 4.0.2)                        
#>  CRAN (R 4.0.2)                        
#> 
#> [1] /home/msheykhmousa/R/x86_64-pc-linux-gnu-library/4.0
#> [2] /usr/lib/R/site-library
#> [3] /usr/lib/R/library
```

</details>



*So the question is now :
- How to solve missing values problems?
- Where and how to use `future.apply` ?
- why I get the error at the first place?

I appreciate any help here.
Best, Mo

**Comment by mllg** [comment]
Looks like this has been an issue in future which is now fixed...
